### PR TITLE
list: show cdcasasagi-managed MCP servers as `name : url`

### LINKED ISSUE #38: Add `list` command to show cdcasasagi-managed MCP servers
## Context

cdcasasagi currently lets users add (`add`, `import`), validate (`validate-import`), revert, and check (`doctor`) the Claude Desktop MCP server config, but offers no way to *see* what is currently configured. Today a user has to open the JSON config file by hand to remember which mcp-proxy-backed servers they previously registered.

Add a `list` command that prints, one per line, the cdcasasagi-managed MCP server entries in the current Claude Desktop config in `name : url` form. This makes the tool's lifecycle (`add` → `list` → `revert`) self-contained.

## Audience & shape

This is a developer-facing read-only command (similar in spirit to `doctor` and `validate-import`):

- No `--write` (read-only, never mutates the config).
- Always safe to run; same exit-code conventions as `doctor` (`0` on success, `1` only on real errors like a corrupt config).

## Behavior

```
$ cdcasasagi list
Target: /Users/you/Library/Application Support/Claude/claude_desktop_config.json

  notion               : https://mcp.notion.com/mcp
  openai-developer-docs : https://developers.openai.com/mcp
```

(Single space around `:`, names left-justified to the longest name so URLs align — matches the `validate_ok_message` style in `src/cdcasasagi/output.py`.)

Empty case:

```
$ cdcasasagi list
No mcp-proxy MCP servers configured.
Target: /Users/you/Library/Application Support/Claude/claude_desktop_config.json
```

### Filtering rule

Show only entries whose `command` resolves to an `mcp-proxy` binary. Use **basename matching**: `Path(entry["command"]).name in {"mcp-proxy", "mcp-proxy.exe"}`.

Rationale: strict matching against `mcp_proxy.resolve_path()` would (a) require mcp-proxy to currently be importable (forcing a hard error for an otherwise read-only command) and (b) hide entries written by a previous cdcasasagi install in a different venv. Basename matching matches the user's intent — "MCP servers that go through mcp-proxy" — without those fragility issues.

Entries with malformed `args` (missing or fewer than 3 elements, or `args[0] != "--transport"`) are skipped silently — they cannot have been written by cdcasasagi's `build_entry`. The URL is taken from `args[-1]`, consistent with `find_entry_names_by_url`.

### Sort order

Alphabetical by name. Predictable, stable across runs, matches what a user scanning a list expects.

### Exit codes

- `0` — listed successfully (including empty list).
- `1` — config JSON failed to parse (`ConfigError`).

### `src/cdcasasagi/cli.py`

Add a new command function below `doctor`:

```python
@app.command(name="list")
def list_cmd() -> None:
    """List cdcasasagi-managed MCP servers as 'name : url'."""
    cfg_path = desktop_config.config_path()
    try:
        config = desktop_config.load_config(cfg_path)
    except desktop_config.ConfigError as e:
        typer.echo(str(e), err=True)
        raise typer.Exit(code=1)

    servers = desktop_config.list_mcp_proxy_entries(config)
    typer.echo(output.list_message(cfg_path, servers))
```

`name="list"` is required because `list` is a Python builtin — same trick already used for `import_cmd` and `validate_import`.

### `src/cdcasasagi/desktop_config.py`

Add a small pure helper next to `find_entry_names_by_url`:

```python
def list_mcp_proxy_entries(config: dict[str, Any]) -> list[tuple[str, str]]:
    """Return ``(name, url)`` for every entry whose command basename is mcp-proxy.

    Entries with unexpected ``args`` shape are skipped.
    """
    result: list[tuple[str, str]] = []
    for name, entry in config.get("mcpServers", {}).items():
        cmd = entry.get("command", "")
        if Path(cmd).name not in {"mcp-proxy", "mcp-proxy.exe"}:
            continue
        args = entry.get("args", [])
        if len(args) < 3 or args[0] != "--transport":
            continue
        result.append((name, args[-1]))
    result.sort(key=lambda x: x[0])
    return result
```

Keeping the logic in `desktop_config.py` (not `cli.py`) follows the project's separation: config-shape knowledge stays out of the CLI layer.

### `src/cdcasasagi/output.py`

Add a formatter, mirroring `validate_ok_message`:

```python
def list_message(config_path: Path, servers: list[tuple[str, str]]) -> str:
    if not servers:
        return (
            "No mcp-proxy MCP servers configured.\n"
            f"Target: {config_path}"
        )
    max_name = max(len(n) for n, _ in servers)
    lines = [f"Target: {config_path}", ""]
    for name, url in servers:
        lines.append(f"  {name.ljust(max_name)} : {url}")
    return "\n".join(lines)
```

### `tests/test_cli.py`

Add `class TestList` with cases:

- `test_lists_entries` — config has two cdcasasagi entries; output contains both `name : url` lines, sorted alphabetically; exit 0.
- `test_empty_servers` — `{"mcpServers": {}}`; exit 0; "No mcp-proxy MCP servers configured." in output.
- `test_no_config_file` — config file absent; exit 0; same empty message (since `load_config` returns `{"mcpServers": {}}`).
- `test_filters_non_mcp_proxy` — config has one mcp-proxy entry and one entry with a different `command`; only the mcp-proxy one appears.
- `test_skips_malformed_args` — entry with `command` matching mcp-proxy but `args` empty / missing `--transport`; silently skipped.
- `test_corrupt_config` — config file contains invalid JSON; exit 1; error to stderr.

Reuse the existing `config_env` fixture — its `tmp_path / "bin" / "mcp-proxy"` is exactly what basename matching expects.

### `tests/test_desktop_config.py`

Add a small unit-test class for `list_mcp_proxy_entries` covering: empty config, basename matching (positive + negative), `mcp-proxy.exe` matching, malformed args skipped, alphabetical sort.

### `tests/test_output.py`

Add tests for `list_message`: empty case message, formatted case (alignment + colon + URL).

### `README.md`

Add a `### list` subsection (between `### import` and `### validate-import`):

```markdown
### list

Show cdcasasagi-managed MCP servers in the config:

    cdcasasagi list

Output is `name : url`, one entry per line, sorted by name. Entries whose command is not `mcp-proxy` are not shown.
```

## Verification

```bash
uv run --no-sync ruff format src/ tests/
uv run --no-sync ruff check --fix src/ tests/
uv run --no-sync pytest

# Manual smoke test
export CLAUDE_DESKTOP_CONFIG=/tmp/list-test.json
echo '{"mcpServers": {"notion": {"command": "/path/to/mcp-proxy", "args": ["--transport", "streamablehttp", "https://mcp.notion.com/mcp"]}}}' > "$CLAUDE_DESKTOP_CONFIG"
uv run --no-sync cdcasasagi list
```

### PR DESCRIPTION
**Pull request by ftnext**
Closes #38.

## Summary

Adds a read-only `cdcasasagi list` command that prints the cdcasasagi-managed MCP server entries in the current Claude Desktop config, one per line as `name : url` (sorted alphabetically by name).

```
$ cdcasasagi list
Target: /Users/you/Library/Application Support/Claude/claude_desktop_config.json

  notion                : https://mcp.notion.com/mcp
  openai-developer-docs : https://developers.openai.com/mcp
```

Empty case:

```
$ cdcasasagi list
No mcp-proxy MCP servers configured.
Target: /Users/you/Library/Application Support/Claude/claude_desktop_config.json
```

## Design notes

- **Filter rule** — basename match: `Path(entry["command"]).name in {"mcp-proxy", "mcp-proxy.exe"}`. Strict matching against `mcp_proxy.resolve_path()` would (a) make this read-only command hard-fail when mcp-proxy is missing and (b) hide entries written by a previous cdcasasagi install in a different venv.
- **Args robustness** — entries with malformed `args` (fewer than 3 elements, or `args[0] != "--transport"`) are skipped silently. URL is taken from `args[-1]`, consistent with `find_entry_names_by_url`.
- **No `--write`** — the command never mutates the config; same shape as `doctor` / `validate-import`.
- **Layered like the rest** — config-shape logic lives in `desktop_config.list_mcp_proxy_entries`; formatting in `output.list_message`; the CLI handler is a thin glue function.

## Files changed

- `src/cdcasasagi/cli.py` — new `list_cmd` (registered as `name="list"` to avoid the Python builtin clash)
- `src/cdcasasagi/desktop_config.py` — new `list_mcp_proxy_entries()`
- `src/cdcasasagi/output.py` — new `list_message()`
- `tests/test_cli.py`, `tests/test_desktop_config.py`, `tests/test_output.py` — coverage for happy path, empty config, missing config file, non-mcp-proxy filtering, malformed-args skipping, corrupt-config error, alignment, ASCII-only output
- `README.md` — `### list` subsection

## Test plan

- [x] `uv run --no-sync ruff format src/ tests/`
- [x] `uv run --no-sync ruff check --fix src/ tests/`
- [x] `uv run --no-sync pytest` — 202 passed
- [x] Manual smoke: empty case, populated case, `--help`

### CODE REVIEW

File: src/cdcasasagi/desktop_config.py [hunk 1]
```diff
@@ -88,6 +88,24 @@ def find_entry_names_by_url(config: dict[str, Any], url: str) -> list[str]:
...
 
+def list_mcp_proxy_entries(config: dict[str, Any]) -> list[tuple[str, str]]:
+    """Return ``(name, url)`` for every entry whose command basename is mcp-proxy.
+
+    Entries with unexpected ``args`` shape are skipped.
+    """
+    result: list[tuple[str, str]] = []
+    for name, entry in config.get("mcpServers", {}).items():
+        cmd = entry.get("command", "")
+        if Path(cmd).name not in {"mcp-proxy", "mcp-proxy.exe"}:
```
[the remark this answers is not in the document]
Comment (ftnext) [reply src/cdcasasagi/desktop_config.py]: Claude Desktopが動いているはずなので、malformedは考慮不要ではないかと思いました

[the remark this answers is not in the document]
Comment (ftnext) [reply src/cdcasasagi/desktop_config.py]: ご指摘ありがとうございます。AGENTS.md の "Assume config structure is valid (Claude Desktop is using it)" に沿うべきところを、Codex の指摘に流されて防御的な型チェックを足してしまっていました。`isinstance(cmd, str)` / `isinstance(args, list)` のガードを 2ebf926 で revert しました。

---
_Generated by [Claude Code](https://claude.ai/code)_
